ls1intum/Artemis
# [Feature] Add impact warning on feedback blocks for textual exercises

### Checklist

### Motivation and Context


This feature allows text exercise graders (tutors) to view the impact of their grading on the other submissions. We expect that having the graders know more about how their added feedback might affect other submissions in terms of how many automatic feedbacks are generated, will make the grading process more qualitative and also more efficient. That's because graders can become more conscious about the effects of their grading and provide better feedback but also as they are grading similar submissions they will encounter more of their own feedback, which means reducing their own grading time as well.

### Description

On the server side:
- A field named `numberOfAffectedSubmissions` is added to the **TextBlock** entity as a non-persistent value. The end goal is to calculate and set this value on the fly and return it to the client since adding a new column was seen to be inefficient. 
- A new query function `getNumberOfOtherBlocksInCluster` is added on the **TextBlockRepository**. The query takes the block id as an input, finds its respective cluster and counts and returns how many other blocks exist in that cluster. As elements in a cluster represent blocks which have a high degree of similarity between each other (=> distanceBetweenBlocks < 1 = DISTANCE_THRESHOLD) the number of elements in a block should represent an approximate number of how many other submissions would receive an AUTOMATIC Feedback label - if the feedback would be submitted. 
-  A new function : `setNumberOfAffectedSubmissionsPerBlock` is added on the **TextBlockService** which:
     - iterates over the blocks of a given submission,
     - for each block uses the query `getNumberOfOtherBlocksInCluster` to find and set the block's `numberOfAffectedSubmissions` field value.
- A check is added on the **prepareSubmissionForAssessment** function which checks whether Automatic Feedback Suggestions are enabled and then calls **setNumberOfAffectedSubmissionsPerBlock** on the referenced submission being prepared. The value is thereafter added to the the TextBlock entity which is then returned to the client through the _/submission/{submissionId}_ endpoint 

On the client side: 
- A card-header element with a warning icon and a text is added on **textblock-feedback-editor.component.html**. The text contains a translated warning sentence with a variable number.  This number is directly fetched from the received TextBlock's `numberOfAffectedSubmissions` field. In case this number is 0 or the feedback has 0 credits then the warning is not shown.
- A field `numberOfAffectedSubmissions` is added on **text-block.model.ts** mirroring the same added field on the server side. 
- Translations for the warning sentence are also added in the respective .json files


### Steps for Testing


**1.** Log in to Artemis on **TS1** as a Tutor/Instructor.

**2.** There are already created TextExercises for the purpose of testing this PR. Text Exercise 7729 in course named ndricim (WS22/23) already has pre-calculated text clusters, meaning you can open and test right away.


**3.** Now go to that TextExercise and pick a submission to grade. Select a text block as shown in the next step and increase the number of points. You do not need to submit, simply having the number of points != 0 should enable the warning. 

**4.** The warning shown in the feedback editor **should** reflect the number of other **textblocks** in the same cluster. In this particular exercise you should be getting the number 5 or4. For simplicity this Text Exercise has only 2 different Text Blocks, so while grading only **two different numbers** of affected submissions should be shown as below 5 or 4:

https://user-images.githubusercontent.com/18335489/120012863-58e64e80-bfe0-11eb-9079-66b2a7f4dc1b.mp4


https://user-images.githubusercontent.com/18335489/120012973-803d1b80-bfe0-11eb-88dc-4907882073f0.mp4



### Test Coverage



New server added function and component tested in 
Coverage: 
- TextBlock.java - 97%
- TextBlockService  - 96%
- TextAssessmentService.java - 93%


### Screenshots

![image](https://user-images.githubusercontent.com/18335489/118553787-13648e80-b761-11eb-9ce6-440cdb1e9b51.png)
